From: Frank Sorenson <sorenson@redhat.com>
Date: Thu, 29 Sep 2016 10:44:38 -0500
Subject: [PATCH] sunrpc: add generic_auth hash_cred() function
Git-commit: 18028c967e423673a055a30b1bf4e603ec64026d
Patch-mainline: v4.9
References: bsc#1012917

Add a hash_cred() function for generic_auth, using both the
uid and gid from the auth_cred.

Signed-off-by: Frank Sorenson <sorenson@redhat.com>
Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>
Acked-by: NeilBrown <neilb@suse.com>

---
 net/sunrpc/auth_generic.c |    9 +++++++++
 1 file changed, 9 insertions(+)

--- a/net/sunrpc/auth_generic.c
+++ b/net/sunrpc/auth_generic.c
@@ -13,6 +13,7 @@
 #include <linux/sunrpc/clnt.h>
 #include <linux/sunrpc/debug.h>
 #include <linux/sunrpc/sched.h>
+#include <linux/hash.h>
 
 #ifdef RPC_DEBUG
 # define RPCDBG_FACILITY	RPCDBG_AUTH
@@ -69,6 +70,13 @@ static struct rpc_cred *generic_bind_cre
 	return auth->au_ops->lookup_cred(auth, acred, lookupflags);
 }
 
+static int
+generic_hash_cred(struct auth_cred *acred, unsigned int hashbits)
+{
+	return hash_64(acred->gid | ((u64)acred->uid << (sizeof(gid_t) * 8)),
+		       hashbits);
+}
+
 /*
  * Lookup generic creds for current process
  */
@@ -172,6 +180,7 @@ void rpc_destroy_generic_auth(void)
 static const struct rpc_authops generic_auth_ops = {
 	.owner = THIS_MODULE,
 	.au_name = "Generic",
+	.hash_cred = generic_hash_cred,
 	.lookup_cred = generic_lookup_cred,
 	.crcreate = generic_create_cred,
 };
